Feature/components button types

MoZarea opened the pull request:
**This commit introduces several new button components and updates existing ones:**

- **NegativeButton** and **NegativeTextButton**: New components for negative actions, utilizing error colors from the theme.
- **ButtonDefaults**: Updated `defaultHeight` to 56.dp and added `defaultFabSize` as 64.dp.
- **FabButton**:
    - Now uses a horizontal gradient from `TudeeTheme.color.primaryGradient` for its background.
    - Hides content when in the `LOADING` state.
- **DefaultButton**:
    - Updated sizing: FABs now use `ButtonDefaults.defaultFabSize` for both width and height, while other buttons use `ButtonDefaults.defaultHeight`.
    - Background gradient is now only applied when the button state is `IDLE` or `LOADING`.
    - Added an `AllButtonPreview` to display all button types and their states.
- **PrimaryButton**: Now uses a horizontal gradient from `TudeeTheme.color.primaryGradient` for its background.

All new and updated button components include comprehensive preview functions to showcase their different states (idle, loading, disabled, error).

![image](https://github.com/user-attachments/assets/23fe18ad-0d02-4ea4-989d-9bb3eafa42c0)

MoZarea committed:
Add composable for spinning progress indicator lines

MoZarea committed:
Add custom button composables

abdallahFc commented on app/src/main/java/com/example/tudee/presentation/composables/buttons/ButtonColors.kt:
```diff
@@ -0,0 +1,89 @@
+    val backgroundColor: Color? = null,
+    val borderColor: Color? = null,
+    val backgroundBrush: Brush? = null
+) {
```
I recommend that ButtonColors contain all the colors, similar to buttonColors in Material 3. Pass all the colors and states to DefaultButton, and let DefaultButton decide which color to show based on the state. Also, remove the companion object

abdallahFc commented on app/src/main/java/com/example/tudee/presentation/composables/buttons/ButtonDefaults.kt:
```diff
@@ -0,0 +1,22 @@
+    val ErrorBackground @Composable get() = TudeeTheme.color.statusColors.errorVariant
+    val ErrorContent @Composable get() = TudeeTheme.color.statusColors.error
+    val Stroke @Composable get() = TudeeTheme.color.textColors.stroke
+    val BorderColor @Composable get() = TudeeTheme.color.textColors.disable
```
why need all this here?

abdallahFc commented on app/src/main/java/com/example/tudee/presentation/composables/buttons/ButtonTypes.kt:
```diff
@@ -0,0 +1,7 @@
+package com.example.tudee.presentation.composables.buttons
+
+
```
remove extra spaces

abdallahFc replied:
Also, why do we need the ButtonType enum if you’ve already created a separate component for each type?

abdallahFc commented on app/src/main/java/com/example/tudee/presentation/composables/buttons/DefaultButton.kt:
```diff
@@ -0,0 +1,73 @@
+    modifier: Modifier = Modifier.Companion,
+    enabled: Boolean = true,
+    state: ButtonState = ButtonState.IDLE,
+    type: ButtonType = ButtonType.PRIMARY,
```
DefaultButton is a high level module and should not depend on lower level , such as specific button types that use it. For example, if we need to add a rounded button in the future, we should be able to use DefaultButton to build it without making any changes to DefaultButton itself.

abdallahFc reviewed (dismissed):
General Notes:
-Add a preview for every component, including all of its states.
-If you need to create an abstraction layer for a complex component, you must reduce the number of parameters passed to it. If you're passing all the same parameters as the parent component, there's no point in creating another layer.

MoZarea committed:
Refactor: Update button components

Key changes include:
- Renamed `ButtonDefaults.DefaultHeight` to `ButtonDefaults.defaultHeight`.
- Renamed `ButtonDefaults.DefaultShape` to `ButtonDefaults.defaultShape`.
- Renamed `ButtonDefaults.DefaultPadding` to `ButtonDefaults.defaultPadding`.
- Introduced a new `ButtonColors` data class within `ButtonDefaults` to manage background color, content color, and background gradient.
- Removed the separate `ButtonColors.kt` and `ButtonTypes.kt` files.
- Updated individual button composables (`PrimaryButton`, `SecondaryButton`, `TextButton`, `FabButton`) to use the new `ButtonColors` and simplified logic for determining colors based on `ButtonState`.
- Added preview composables for each button type in their respective files to showcase different states (idle, loading, disabled, error).
- `DefaultButton` now handles background and border modifiers more directly using `clip`, `background`, and `border` modifiers.
- Simplified `DefaultButton` by removing the `type` parameter and adjusting the progress indicator display logic.

Rodinamomen commented on app/src/main/java/com/example/tudee/presentation/composables/buttons/ButtonDefaults.kt:
```diff
@@ -0,0 +1,27 @@
+    val contentColor: Color= Color.Unspecified,
+    val backgroundGradient: Brush? =null ,
+)
+object ButtonDefaults {
```
You can add all of buttons withing one file that has more than one function

Rodinamomen commented on app/src/main/java/com/example/tudee/presentation/composables/buttons/DefaultButton.kt:
```diff
@@ -0,0 +1,117 @@
+import com.example.tudee.presentation.composables.SpinningProgressIndicatorLines
+
+@Composable
+fun DefaultButton(
```
Support button text

Rodinamomen commented on app/src/main/java/com/example/tudee/presentation/composables/buttons/DefaultButton.kt:
```diff
@@ -0,0 +1,117 @@
+        modifier = modifier
+            .height(ButtonDefaults.defaultHeight)
+            .clip(shape)
+            .then(if(colors.backgroundGradient!=null) Modifier.background(colors.backgroundGradient,shape) else Modifier.background(colors.backgroundColor,shape))
```
ou can set color to be deafult and if the developer needs to change it he can 
instead of bakcing background color nullable

Rodinamomen commented on app/src/main/java/com/example/tudee/presentation/composables/buttons/DefaultButton.kt:
```diff
@@ -0,0 +1,117 @@
+            .height(ButtonDefaults.defaultHeight)
+            .clip(shape)
+            .then(if(colors.backgroundGradient!=null) Modifier.background(colors.backgroundGradient,shape) else Modifier.background(colors.backgroundColor,shape))
+            .then(if (border != null) Modifier.border(border, shape) else Modifier)
```
Null can be used if the composable will appear or not 
default values when I supprot custome color and developer can change it

Rodinamomen commented on app/src/main/java/com/example/tudee/presentation/composables/buttons/FabButton.kt:
```diff
@@ -0,0 +1,87 @@
+    shape: Shape = ButtonDefaults.defaultShape,
+    contentPadding: PaddingValues = ButtonDefaults.defaultPadding,
+    buttonColors: ButtonColors = ButtonDefaults.colors(),
+    content: @Composable RowScope.() -> Unit
```
here you should support the content for each type not the developer how made it

Rodinamomen commented on app/src/main/java/com/example/tudee/presentation/composables/buttons/PrimaryButton.kt:
```diff
@@ -0,0 +1,90 @@
+    shape: Shape = ButtonDefaults.defaultShape,
+    contentPadding: PaddingValues = ButtonDefaults.defaultPadding,
+    buttonColors: ButtonColors = ButtonDefaults.colors(),
+    content: @Composable RowScope.() -> Unit
```
if it will always be text and near it an icon make them with the text color every thing related to it and and user just sends text as string res

EmanAbobakr commented on app/src/main/java/com/example/tudee/presentation/composables/SpinningProgressIndicatorLines.kt:
```diff
@@ -0,0 +1,85 @@
+        ),
+        label = "progress"
+    )
+    ButtonDefaults
```
why do you use ButtonDefaults here?

EmanAbobakr commented on app/src/main/java/com/example/tudee/presentation/composables/SpinningProgressIndicatorLines.kt:
```diff
@@ -0,0 +1,85 @@
+
+            val currentActiveIndex = (animationProgress * lines).toInt() % lines
+            val distanceFromActive = kotlin.comparisons.minOf(
+                abs(i - currentActiveIndex),
```
you can use minOf without kotlin.comparisons.minOf

EmanAbobakr commented on app/src/main/java/com/example/tudee/presentation/composables/buttons/ButtonDefaults.kt:
```diff
@@ -0,0 +1,27 @@
+data class ButtonColors(
+    val backgroundColor: Color = Color.Unspecified,
+    val contentColor: Color= Color.Unspecified,
+    val backgroundGradient: Brush? =null ,
```
make sure to format the files before pushing

EmanAbobakr commented on app/src/main/java/com/example/tudee/presentation/composables/buttons/DefaultButton.kt:
```diff
@@ -0,0 +1,117 @@
+import androidx.compose.ui.graphics.Shape
+import androidx.compose.ui.tooling.preview.Preview
+import androidx.compose.ui.unit.dp
+import com.example.tudee.designsystem.theme.TudeeTheme
```
remove unused imports plz

EmanAbobakr commented on app/src/main/java/com/example/tudee/presentation/composables/buttons/DefaultButton.kt:
```diff
@@ -0,0 +1,117 @@
+            .clip(shape)
+            .then(if(colors.backgroundGradient!=null) Modifier.background(colors.backgroundGradient,shape) else Modifier.background(colors.backgroundColor,shape))
+            .then(if (border != null) Modifier.border(border, shape) else Modifier)
+            .clickable(enabled = enabled) { onClick() },
```
you are using enabled here for only the click functionality, it won't be grayed button as Figma

EmanAbobakr commented on app/src/main/java/com/example/tudee/presentation/composables/buttons/TextButton.kt:
```diff
@@ -0,0 +1,100 @@
+import androidx.compose.ui.graphics.Shape
+import androidx.compose.ui.tooling.preview.Preview
+import androidx.compose.ui.unit.dp
+import com.example.tudee.designsystem.theme.TudeeTheme
```
remove unused imports plz

Mohamed-Zarea committed:
Refactor: Update button composables

This commit refactors the button composables and their default values.

Key changes:
- Updated `ButtonDefaults` to include `disabledBackgroundColor`, `disabledContentColor`, `errorBackgroundColor`, and `errorContentColor`.
- Increased `defaultHeight` and added `defaultFabPadding` in `ButtonDefaults`.
- Modified `DefaultButton` to handle different button states (IDLE, LOADING, DISABLED, ERROR) for `contentColor` and `backgroundColor`.
- Added `isFab` parameter to `DefaultButton` to adjust padding and width for FABs.
- Updated `FabButton`, `SecondaryButton`, and `TextButton` to utilize the new state handling in `DefaultButton` and the updated `ButtonDefaults`.
- Adjusted `SpinningProgressIndicatorLines` to use `minOf` and removed an unused import.
- Updated previews for all button types.

Mohamed-Zarea committed:
Merge branch 'develop' into feature/components-button-types

Mohamed-Zarea committed:
fix conflict: Update button default colors

The `disabledBackgroundColor` and `disabledContentColor` in `ButtonDefaults`, `SecondaryButton`, and `TextButton` were updated to use `TudeeTheme.color.disable` and `TudeeTheme.color.stroke` respectively, removing the `textColors` intermediate path.

Mohamed-Zarea committed:
feat: implement NegativeButton & Negative Text Button components

Mohamed-Zarea committed:
feat: implement NegativeButton & Negative Text Button components

mohamedshemees commented on app/src/main/java/com/example/tudee/presentation/composables/buttons/FabButton.kt:
```diff
@@ -0,0 +1,96 @@
+    buttonColors: ButtonColors = ButtonDefaults.colors(),
+    content: @Composable RowScope.() -> Unit
+) {
+    val currentContent = if (state == ButtonState.LOADING) {{}}else content
```
are the double brackets mandatory ?

MoZarea replied:
![image](https://github.com/user-attachments/assets/bd94f9e6-bd45-4e5d-b1ba-62e631466421)
if i did make it doubled it will be Any
i
if u have any other solution please tell me about

EmanAbobakr commented on app/src/main/java/com/example/tudee/presentation/composables/buttons/FabButton.kt:
```diff
@@ -15,33 +17,18 @@ fun FabButton(
         modifier = modifier,
         state = state,
         enabled = enabled,
+        isFab = true,
```
the parameter name is confusing, could we choose a more clear name?

MoZarea replied:
which param?

EmanAbobakr replied:
isFab it's not clear what it is used for, you even don't have a doc for the function to make it clearer

Mohamed-Zarea committed:
Merge remote-tracking branch 'origin/feature/components-button-types' into feature/components-button-types

# Conflicts:
#	app/src/main/java/com/example/tudee/presentation/composables/buttons/DefaultButton.kt

Mohamed-Zarea committed:
fix: rename isFab parameter in DefaultButton Composable to be isFabType
